src/sql/reuse/relaxed.rs

```diff
@@ -0,0 +1,147 @@
+        }
+
+        // Checks group by compatibility between queries.
+        for (srcdst, ex_qge) in &existing_qg.edges {
+            match *ex_qge {
+                QueryGraphEdge::GroupBy(ref ex_columns) => {
+                    if !new_qg.edges.contains_key(srcdst) { return Some(ReuseType::PrefixReuse); }
+                    let new_qge = &new_qg.edges[srcdst];
+                    match *new_qge {
+                        QueryGraphEdge::GroupBy(ref new_columns) => {
+                            // GroupBy implication holds if the new QG groups by the same columns as
+                            // the original one, or by a *superset* (as we can always apply more
+                            // grouped operatinos on top of earlier ones)
+                            if new_columns.len() < ex_columns.len() {
+                                // more columns in existing QG's GroupBy, so we're done
+                                // however, we can still reuse joins and predicates.
+                                return Some(ReuseType::PrefixReuse);
+                            }
+                            for ex_col in ex_columns {
+                                // EQG groups by a column that we don't group by, so we can't reuse
+                                // the group by nodes, but we can still reuse joins and predicates.
+                                if !new_columns.contains(ex_col) {
+                                    return Some(ReuseType::PrefixReuse);
+                                }
+                            }
+                        }
+                        // If there is no matching GroupBy edge, we cannot reuse the group by clause
+                        _ => return Some(ReuseType::PrefixReuse),
+                    }
+                }
+                _ => continue
+            }
+        }
+
+        // Check that the new query's predicates imply the existing query's predicate.
+        for (name, ex_qgn) in &existing_qg.relations {
+            if !new_qg.relations.contains_key(name) { return Some(ReuseType::PrefixReuse); }
+            let new_qgn = &new_qg.relations[name];
+
+            // iterate over predicates and ensure that each matching one on the existing QG is implied
+            // by the new one
+            for ep in &ex_qgn.predicates {
+                let mut matched = false;
+
+                for np in &new_qgn.predicates {
+                    if complex_predicate_implies(np, ep) {
+                        matched = true;
+                        break
+                    }
+                }
+                if !matched {
+                    // We found no matching predicate for np, so we give up now.
+                    // However, we can still reuse the join nodes from the existing query.
+                    return Some(ReuseType::PrefixReuse);
+                }
+            }
+        }
+
+        // we don't need to check projected columns to reuse a prefix of the query
```

Comment:
This comment might need updating in light of the fact that we use "prefix" differently now ;)